### ISSUE TITLE
PSReadLine redraws entire line with each key press 

### ISSUE THREAD

**Issue by ghost** [PowerShell/PSReadLine#420]
On Windows, with a Narrator enabled (often used by visually empaired users), user opens Powershell
When you type a character there are two wrong results depending on the narrator in use
Expected: The new character is spoken and when a word is complete the word is spoken
    Actual using NVDA: no text is spoken
    Actual using Narrator: the entire line is spoken
Requested change is for PSReadline to not redraw the entire line every time a character is typed when a Navigator is in use.

**Comment by cspotcode** [comment]
@lzybkr Can you point me towards the relevant code?  I'd like to attempt a fix for this.

If the code is currently redrawing the entire line, it should be possible to amend that logic without too much added complexity.

Step through the line in memory and compare to the previous line, stopping at the first difference.  Redrawing can be done starting from that point.  So if the only change was appending a character (which should usually be the case) then the cursor will jump to its current location and draw a single character.

Does that make sense?  Is "render.cs", `private void Render` where I should look to get started?

**Comment by lzybkr** [comment]
Thanks for the offer @cspotcode. This issue has been improved significantly since it was opened, but I left it open because improvements are still possible. That said, I'm not sure if improvements will be noticeable, maybe on a really slow terminal.

Most rendering is done in `render.cs`, and the current code tracks what was previously written. The trickiest bit will be maintaining correct colors. For example, if the user typed:

```
PS> Get-
```

Then typed `C`, you can't just write the `C`, you'd need to use the last VT sequence in addition to writing the `C`.

There is also some rendering code in `completions.cs`, but I think that code only updates what has changed.

**Comment by cspotcode** [comment]
@lzybkr thanks for the quick reply!  This issue makes PSReadline pretty uncomfortable inside WSL.  I can see the cursor flickering all over the place and sometimes it even places floating characters off to the right and triggers the buffer to scroll up when it shouldn't.

For example, here's what I see when I open PowerShell on Windows 10, run `bash -c pwsh`, type a few commands to get the buffer to scroll to the bottom of the screen, and then hold down the `a` key.

![image](https://user-images.githubusercontent.com/376504/40853132-f7489df6-659a-11e8-9e1e-cff8e27892d1.png)

I could be wrong, but I'm assuming this behavior can be improved if PSReadline doesn't redraw the entire line.

**Comment by lzybkr** [comment]
Yeah - the version shipped with `pwsh` is not great. My new code in this repo is much better. `2.0.0-beta1` is available in the gallery and works great with any version of PowerShell >= 3.0.

**Comment by cspotcode** [comment]
@lzybkr I should be running 2.0.0 already.  I installed it with `Install-Module PSReadline -RequiredVersion 2.0.0 -Scope CurrentUser`; is there something else I need to configure?  Is pwsh forcing itself to use the older, bundled version?

```
PS /mnt/c/Users/abradley>       get-module psreadline                                                                   
ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     2.0.0      PSReadLine                          {Get-PSReadlineKeyHandler, Get-PSReadlineOption, Remove-PS...

PS /mnt/c/Users/abradley>
```

**Comment by lzybkr** [comment]
It does look like you're using 2.0.0.

The cursor shouldn't flicker, it's supposed to be hidden during rendering, and that definitely works with Windows PowerShell as well as PowerShell on Windows and Linux, I haven't tried Mac or WSL.
